src/eal.cpp

```diff
@@ -80,34 +83,89 @@ namespace eal {
+  vector<double> getRotation(vector<vector<double>> rotations,
                              vector<double> times, double time,  interpolation interp) {
-    vector<double> coordinate = {0.0, 0.0, 0.0};
+    // Check that all of the data sizes are okay
+    // TODO is there a cleaner way to do this? We're going to have to do this a lot.
+    if (rotations.size() != 4) {
+     throw invalid_argument("Invalid input rotations, expected four vectors.");
+    }
+
+    // Alot of copying and reassignment becuase conflicting data types
+    // probably should rethink our vector situation to guarentee contiguous
+    // memory. Should be easy to switch to a contiguous column-major format
+    // if we stick with Eigen.
+    double data[] = {0,0,0,0};
+    for (size_t i = 0; i<rotations[0].size(); i++) {
+      Eigen::Quaterniond quat(rotations[0][i], rotations[1][i], rotations[2][i], rotations[3][i]);
+      quat.normalize();
+      rotations[0][i] = quat.w();
+      rotations[1][i] = quat.x();
+      rotations[2][i] = quat.y();
+      rotations[3][i] = quat.z();
+    }
+
+    // GSL setup
+    vector<double> coordinate = {0.0, 0.0, 0.0, 0.0};
+
+    coordinate = { interpolate(rotations[0], times, time, interp, 0),
+                   interpolate(rotations[1], times, time, interp, 0),
+                   interpolate(rotations[2], times, time, interp, 0),
+                   interpolate(rotations[2], times, time, interp, 0)};
+
+    Eigen::Quaterniond quat(coordinate.data());
+    quat.normalize();
+
     return coordinate;
   }
 
-  vector<double> getAngularVelocity(string from, string to, vector<vector<double>> rotations,
+  vector<double> getAngularVelocity(vector<vector<double>> rotations,
                                     vector<double> times, double time,  interpolation interp) {
-    vector<double> coordinate = {0.0, 0.0, 0.0};
+    // Check that all of the data sizes are okay
+    // TODO is there a cleaner way to do this? We're going to have to do this a lot.
+    if (rotations.size() != 4) {
+     throw invalid_argument("Invalid input rotations, expected four vectors.");
+    }
+
+    double data[] = {0,0,0,0};
+    for (size_t i = 0; i<rotations[0].size(); i++) {
+      Eigen::Quaterniond quat(rotations[0][i], rotations[1][i], rotations[2][i], rotations[3][i]);
+      quat.normalize();
+      rotations[0][i] = quat.w();
+      rotations[1][i] = quat.x();
+      rotations[2][i] = quat.y();
+      rotations[3][i] = quat.z();
+    }
+
+    // GSL setup
+    vector<double> coordinate = {0.0, 0.0, 0.0, 0.0};
+
+    coordinate = { interpolate(rotations[0], times, time, interp, 1),
+                   interpolate(rotations[1], times, time, interp, 1),
+                   interpolate(rotations[2], times, time, interp, 1),
+                   interpolate(rotations[2], times, time, interp, 1)};
+
+    Eigen::Quaterniond quat(coordinate.data());
+    quat.normalize();
```

Comment:
Does this modify the internal data array of coordinate?

I'd expect the Eigen quaternion constructor to copy the contents of the array.

Reply:
If I understand he docs, it uses the array as the internal buffer, so it should work with the pointer directly instead of copying it. It's why I suggested changing our vector format, we can stuff a std::vector into an eigen::Matrix with only copying a pointer. In all fairness I need to test it first since I am only going off what I read.

Reply:
Needed to use Eigen::Map to get this functionality, why only copying the pointer isn't the default behavior for Eigen matrices I don't know.